src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -69,6 +69,76 @@ protected override ImmutableArray<LocalSymbol> BuildLocals()
             }
         }
 
+        internal void CollectLocalsFromDeconstruction(
```

Comment:
Yes, I had gone down that route and then reverted because it only re-used a small part of ExpressionVariableFinder and also had a problem of context. It seems it's ok to check parents during visit, and that will help to stick with the visitor pattern (no separate "CollectLocalsFromDeconstruction" method). I'll simplify as part of #15064.